wasm-text.md

```diff
@@ -536,41 +545,64 @@ Since we do not have polymorphic functions available, we define one function per
-      => #module( ... types: TS
+      => #module( ... types: #t2aDefns<C>(TS)
                     , funcs: #t2aDefns<C>(FS)
                     , tables: TABS
                     , mems: MS
                     , globals: GS
                     , elem: #t2aDefns<C>(EL)
                     , data: DAT
                     , start: #t2aDefns<C>(S)
-                    , importDefns: IS
+                    , importDefns: #t2aDefns<C>(IS)
                     , exports: #t2aDefns<C>(ES)
                     , metadata: #meta(... id: OID, funcIds: FIDS)
                 )
 ```
 
+
+#### Types
+
+```k
+    rule #t2aDefn<_>((type OID (func TDECLS))) => #type(... type: asFuncType(TDECLS), metadata: OID)
+```
+
+#### Imports
+
+TODO
+```k
+    rule #t2aDefn<ctx(... typeIds: TIDS)>(( import MOD NAME (func OID:OptionalId (type ID:Identifier)            ))) => ( import MOD NAME (func OID:OptionalId (type {TIDS[ID]}:>Int)))
+    rule #t2aDefn<ctx(... typeIds: TIDS)>(( import MOD NAME (func OID:OptionalId (type ID:Identifier) _:TypeDecls))) => ( import MOD NAME (func OID:OptionalId (type {TIDS[ID]}:>Int)))
+```
+
 #### Functions
 
+After unfolding, each type use in a function starts with an explicit reference to a module-level function.
+
 ```k
-    rule #t2aDefn<C>(( func OID:OptionalId T:TypeUse LS:LocalDecls IS:Instrs ))
-      => #func(... type: #t2aTypeUse <#updateLocalIds(C, #ids2Idxs(T, LS))>(T)
-                 , locals: #t2aLocalDecls<#updateLocalIds(C, #ids2Idxs(T, LS))>(LS)
+    rule #t2aDefn<ctx(... typeIds: TIDS) #as C>(( func OID:OptionalId T:TypeUse LS:LocalDecls IS:Instrs ))
+      => #func(... type: typeUse2typeIdx(T, TIDS)
+                 , locals: locals2vectype(LS)
                  , body: #t2aInstrs <#updateLocalIds(C, #ids2Idxs(T, LS))>(IS)
                  , metadata: #meta(... id: OID, localIds: #ids2Idxs(T, LS))
               )
 
-    syntax TypeUse    ::= "#t2aTypeUse"   "<" Context ">" "(" TypeUse    ")" [function]
-    syntax LocalDecl  ::= "#t2aLocalDecl" "<" Context ">" "(" LocalDecl  ")" [function]
- // -----------------------------------------------------------------------------------
-    rule #t2aTypeUse<_>((type TYP) _TDS:TypeDecls      ) => (type TYP)
-    rule #t2aTypeUse<C>((param _ID:Identifier AVT) TDS ) => (param AVT) {#t2aTypeUse<C>(TDS)}:>TypeDecls
-    rule #t2aTypeUse<_>(TU) => TU [owise]
+    syntax Int ::= typeUse2typeIdx ( TypeUse , Map ) [function]
+ // -----------------------------------------------------------
+    rule typeUse2typeIdx( (type ID:Identifier )  ,  TIDS ) => {TIDS [ ID ]}:>Int
+    rule typeUse2typeIdx( (type ID:Identifier ) _,  TIDS ) => {TIDS [ ID ]}:>Int
+    rule typeUse2typeIdx( (type IDX:Int       )  , _TIDS ) => IDX
+    rule typeUse2typeIdx( (type IDX:Int       ) _, _TIDS ) => IDX
```

Comment:
Could also consider writing these as:

```suggestion
    rule typeUse2typeIdx( (type ID:Identifier )  ,  TIDS ) => {TIDS [ ID ]}:>Int
    rule typeUse2typeIdx( (type IDX:Int       )  , _TIDS ) => IDX
    
    rule typeUse2typeIdx( (type ID:Identifier ) _ => (type ID),  TIDS )
    rule typeUse2typeIdx( (type IDX:Int       ) _ => (type IDX), _TIDS )
```

Reply:
Done (with slight variation, last two cases can be merged)